### PR TITLE
MoE routing analysis and metrics capture

### PR DESCRIPTION
**Pull request by mathemakitten**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do ?
* Adds a `RouterTracer` class that captures per-layer top-K routing decisions via forward hooks on TopKRouter modules, writing one JSONL record per (step, layer, rank). Works identically for training and inference.
* Adds `tools/moe_routing/` with two analysis scripts that work on traces from either training or inference:
  - `analyze_routing_concentration.py` — per-layer expert activation frequency, top-N coverage curves, and a heatmap of hot experts vs layers.
  - `analyze_routing_predictability.py` — applies layer L's router weights to L_prev's hidden states and measures cosine/Spearman similarity between predicted and actual per-expert token-count distributions. High scores here are the affirmative signal that distributional routing is predictable one layer ahead.

**Notes on usability**
- Use --moe-routing-trace-max-iters / --moe-routing-trace-max-steps to bound the capture window; the tracer self-disables and removes its hooks after the limit is reached
- Python forward hooks do not fire during CUDA graph replay. Run with `--cuda-graph-impl none` when collecting traces.
- `analyze_routing_predictability.py` requires `--moe-routing-trace-capture-hidden-states` and `--moe-routing-trace-dump-weights` at collection time.

**Follow-up items**
*  Unify hook registration helpers: `megatron/training/activation_logging.py` has a generic `_register_hooks(model, module_types, hook_factory)` utility that walks model chunks and registers forward hooks. `RouterTracer.register_hooks` does the same thing with a loop. These can be consolidated with a shared helper into `core/`, but requires some more work to unify the hook registration pattern (`_register_hooks` in `activation_logging.py` does not currently live in core).

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/moe/router_trace.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+import os
+import threading
+from typing import List, Optional
+
+import torch
+
+_TRACER: Optional["RouterTracer"] = None
+
+
+def init_tracer(
```
Comment (santhnm2) [inline megatron/core/transformer/moe/router_trace.py]: Can we make these function names (`init_tracer` and `get_tracer`) more specific to MoE routing? For example `init_moe_router_tracer` / `get_moe_router_tracer`. Same thing with global `_TRACER`

File: examples/inference/advanced/gpt_dynamic_inference_with_coordinator.py [hunk 2]
```diff
@@ -225,8 +226,23 @@ async def main(
...
+                max_steps=max_steps,
+                rank=rank,
+                capture_hidden_states=getattr(args, 'moe_routing_trace_capture_hidden_states', False),
+                capture_logits=getattr(args, 'moe_routing_trace_capture_logits', False),
+                dump_router_weights=getattr(args, 'moe_routing_trace_dump_weights', False),
+            )
+
         model = get_model_for_inference()
 
+        if get_tracer() is not None:
```
Comment (santhnm2) [inline examples/inference/advanced/gpt_dynamic_inference_with_coordinator.py]: Nit: could use walrus operator here to avoid calling `get_tracer()` twice
```
if (tracer := get_tracer()) is not None:
    tracer.register_hooks(model)
```

File: megatron/core/transformer/moe/router_trace.py [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+    """
+    if "hs_offset" not in record:
+        raise ValueError("Record does not contain hidden-state metadata.")
+    path = os.path.join(trace_dir, f"hidden_states_rank{record['rank']}.bin")
+    with open(path, "rb") as f:
+        f.seek(record["hs_offset"])
+        data = f.read(record["hs_bytes"])
+    arr = torch.frombuffer(bytearray(data), dtype=torch.int16)
+    arr = arr.view(torch.bfloat16).reshape(record["hs_shape"])
+    return arr.clone()
```
Comment (santhnm2) [inline megatron/core/transformer/moe/router_trace.py]: Why is the `clone` necessary? Same thing for `load_logits_for_record`

File: megatron/core/transformer/moe/router_trace.py [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+def _unwrap_model(model):
+    """Unwrap DDP/FSDP wrappers to get the underlying module."""
+    try:
+        from megatron.core.utils import unwrap_model
+
+        return unwrap_model(model)
+    except Exception:
+        while hasattr(model, "module"):
+            model = model.module
+        return model
```
Comment (santhnm2) [inline megatron/core/transformer/moe/router_trace.py]: Why do we need this? Isn't the existing `unwrap_model` sufficient?

File: megatron/inference/utils.py [hunk 5]
```diff
@@ -285,6 +286,15 @@ def add_inference_args(parser: ArgumentParser) -> ArgumentParser:
...
     )
+
+    group.add_argument(
+        "--moe-routing-trace-max-steps",
+        type=int,
+        default=None,
+        help="Maximum number of decode steps to trace (inference). Default is unlimited."
+             "Training uses --moe-routing-trace-max-iters instead.",
+    )
+
```
Comment (santhnm2) [inline megatron/inference/utils.py:326]: Maybe we should make the training / inference argument split more explicit? Otherwise `iters` and `steps` seem pretty close semantically / could be easily confused. So `--moe-routing-trace-max-steps` -> `moe-routing-trace-max-inference-steps` and `--moe-routing-trace-max-iters` -> `--moe-routing-trace-max-training-iters`
Or if `--moe-routing-trace-max-iters` is already in use for training maybe we could just do the inference name change

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/transformer/moe/router_trace.py [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+        dump_router_weights: bool = False,
+    ) -> None:
+        self.output_dir = output_dir
+        self.max_steps = max_steps
+        self.rank = rank
+        self.training_mode = training_mode
+        self.step_id = 0
+        self.layers_seen_this_step: set[int] = set()
+        self.records: list[dict] = []
+        self._lock = threading.Lock()
```
Comment (Mellonta) [inline megatron/core/transformer/moe/router_trace.py]: What's the purpose of the lock here, is there any contention between different threads during inference?

File: megatron/core/transformer/moe/router_trace.py [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+        def hook(module, inputs, outputs):
+            if self._stopped:
+                return
+            if torch.cuda.is_current_stream_capturing():
+                return
+            self._record(module, inputs, outputs, identity)
+
+        return hook
+
+    def _extract_hidden_state(self, inputs, expected_num_tokens):
```
Comment (Mellonta) [inline megatron/core/transformer/moe/router_trace.py:274]: Could you add an one-line comment here explaining what it does?

File: tools/moe_routing/analyze_routing.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    # 1. Expert concentration (hot-set size).
+    run(
+        "analyze_routing_concentration.py",
+        [args.trace_dir] + nexpert_args + outdir_args,
+        "Expert concentration  (hot-set size, routing distribution)",
+    )
+
+    # 2. Distribution predictability: how well do L_prev's hidden states predict L's routing?
+    run(
+        "analyze_routing_predictability.py",
```
Comment (yaox12) [inline tools/moe_routing/analyze_routing.py:86]: **[IMPORTANT Correctness]** The dispatcher calls `analyze_routing_predictability.py`, but that file is not added in this PR.

The documented top-level analysis command will always fail or warn on the second analysis, and the README advertises predictability output that users cannot actually run. Please either add the missing script in this PR, or remove the dispatcher call and README references until the predictability analysis lands.

File: tools/moe_routing/analyze_routing_concentration.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+    print(header)
+    print("  " + "-" * (len(header) - 2))
+
+    per_layer_results = []
+    for layer in layers:
+        freqs = per_layer_freq[layer]
+        ntoken = per_layer_tokens[layer]
+        unique_experts = len(freqs)
+        cov = topk_coverage(freqs, n_values)
+        uniform_bl = 22 / args.num_experts  # random baseline at N=22
```
Comment (yaox12) [inline tools/moe_routing/analyze_routing_concentration.py]: **[IMPORTANT Correctness]** The concentration analysis hard-codes router top-k as `22`.

Models with `moe_router_topk != 22` get incorrect baselines and heatmap rates, even though each trace record already includes `topk`. Please read `topk` from the trace records, or add/pass a `--top-k` argument and use it consistently here and in the heatmap normalization below.

File: tools/moe_routing/analyze_routing_concentration.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    pattern = os.path.join(trace_dir, "router_trace_rank*.jsonl")
+    paths = sorted(glob.glob(pattern))
+    if not paths:
+        raise FileNotFoundError(f"No trace files matching {pattern}")
+    for path in paths:
+        with open(path) as f:
+            for line in f:
+                r = json.loads(line)
+                yield (
+                    r["rank"], r["step"], r["layer"],
```
Comment (yaox12) [inline tools/moe_routing/analyze_routing_concentration.py]: **[IMPORTANT Correctness]** The loader drops `block` and `mtp_idx`, then the analysis aggregates only by integer `layer`.

Decoder layer 0 and MTP layer 0 can collide, even though `router_trace.py` writes `mtp_idx` specifically to avoid that ambiguity. That makes the reported per-layer concentration silently merge distinct routers. Please preserve a full identity such as `(block, mtp_idx, layer)` throughout the counters and output labels.

File: megatron/core/transformer/moe/router_trace.py [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+        """
+        with self._lock:
+            if step_id is not None:
+                for rec in self.records:
+                    rec["step"] = step_id
+                self.step_id = step_id
+            self._flush_records_to_disk()
+            self.step_id += 1
+            self.layers_seen_this_step.clear()
+            if self.step_id >= self.max_steps:
```
Comment (yaox12) [inline megatron/core/transformer/moe/router_trace.py]: **[IMPORTANT Correctness]** `max_steps` is compared against `step_id`, but training passes the absolute iteration number into `advance_step(iteration)`.

On checkpoint resume, `--moe-routing-trace-max-training-iters 500` can stop immediately if the resumed iteration is already >= 500, instead of tracing 500 iterations as documented. Please track a separate captured-step counter, or compute an absolute stop iteration from `args.iteration + moe_routing_trace_max_training_iters`.

File: megatron/core/transformer/moe/router_trace.py [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+                    if self.step_id >= self.max_steps:
+                        self._stopped = True
+                        return
+                self.layers_seen_this_step.add(layer_key)
+
+            # The router-weight dump is keyed by the integer `layer` to match the
+            # JSONL records' `layer` field, which is how downstream analysis joins
+            # weights to traces. (`layers_seen_this_step` above uses the full
+            # `layer_key` only for collision-free step-boundary detection.)
+            if self.dump_router_weights and layer not in self._router_state:
```
Comment (yaox12) [inline megatron/core/transformer/moe/router_trace.py]: **[IMPORTANT Correctness]** Router weight sidecars are keyed only by integer `layer`.

Decoder/MTP routers, or multiple MTP heads with the same inner layer, can overwrite or suppress each other here. A predictability analysis would then join records to the wrong router weights. Please key router state with the same full identity used in records, e.g. `(block, mtp_idx, layer)`, or an equivalent string key.

**Review by yaox12 — COMMENTED** [review]
Post some comments from codex.

**Review by yaox12 — APPROVED** [review]

**Review by Mellonta — APPROVED** [review]

**Review by Phlip79 — CHANGES_REQUESTED** [review]
Can you please add tests?

**Review by Phlip79 — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]

File: megatron/training/config/training_config.py
Comment (maanug-nv) [inline megatron/training/config/training_config.py:1]: wouldn't `LoggerConfig` make more sense for these settings?

**Review by kvareddy — APPROVED** [review]

File: megatron/training/config/training_config.py
Comment (mathemakitten) [reply megatron/training/config/training_config.py:1]: Yes, moved.

**Review by deepakn94 — APPROVED** [review]
